storage_controller/src/tenant_shard.rs

```diff
@@ -646,6 +646,49 @@ impl TenantShard {
         Ok(())
     }
 
+    /// Attempt to reschedule the tenant shard to one of its secondary locations.
+    /// If no schedulable secondary is found, roll back any changes to the intent state.
+    /// If the `attached_to` node argument doesn't match the intent state, no changes
+    /// are made and the function returns successfully.
+    pub(crate) fn reschedule_to_secondary(
+        &mut self,
+        attached_to: NodeId,
+        scheduler: &mut Scheduler,
+        context: &mut ScheduleContext,
+    ) -> Result<(), ScheduleError> {
+        let original_secondaries = self.intent.get_secondary().clone();
+
+        if self.intent.demote_attached(scheduler, attached_to) {
+            let res = match self.schedule(scheduler, context) {
```

Comment:
I'd prefer to call into scheduler.node_preferred(self.intent.get_secondary()).is_some() here to check if there is a usable secondary, rather than mutating the intent and then backing it out if we got a new node.

schedule()'s job is "make it match the scheduling policy", which isn't what we want here: we specifically only want to modify it in one way, so I think it's justified to unpack that: find a usable secondary, and if we have one then do demote_attached,promote_attached .


Might be able to skip the whole call to schedule() at that point: we know exactly

Reply:
> I'd prefer to call into scheduler.node_preferred(self.intent.get_secondary()).is_some() here to check if there is a usable secondary, rather than mutating the intent and then backing it out if we got a new node.

I don't really like the idea of using on scheduling internals like this. We'd be leaking scheduling logic in places where it shouldn't be.

> schedule()'s job is "make it match the scheduling policy", which isn't what we want here

It sort of is though. The way I think about it is: "I've changed your constraints (demoted the draining node), now go figure out how to satisfy those constraints".

Reply:
I've played around with the storage controller scale test and, experimentally, not calling `schedule` in the drain and fill stages results in much more stable placements. This is because secondaries have no chance to flake around. Will close this PR and open a different one informed by the scale testing.

Reply:
Please see https://github.com/neondatabase/neon/pull/8119